for chr in {1..22} X Y
do
echo "java -Xmx64g -jar /data/houl3/Softwares/gatk_3.5/GenomeAnalysisTK.jar -T GenotypeGVCFs -R /data/houl3/BSC_FAM/gatk_bundle/human_g1k_v37_decoy.fasta -nt 32 -L $chr -o BSC_FAM.release1.raw.snps.indels.chr$chr.vcf --dbsnp /data/houl3/BSC_FAM/gatk_bundle/dbsnp_138.b37.vcf --variant Ambigen/Ambigen.raw.snps.indels.chr$chr.union.regions.g.vcf --variant Australian_WGS/Australian.WGS.raw.snps.indels.chr$chr.g.vcf --variant BoMaBa/gVCFs/BoMaBa.raw.snps.indels.chr$chr.union.regions.g.vcf --variant Canadian/Canadian.raw.snps.indels.chr$chr.part1.g.vcf --variant Canadian/Canadian.raw.snps.indels.chr$chr.part2.g.vcf --variant Canadian-Iranian/Canadian-Iranian.raw.snps.indels.chr$chr.union.regions.g.vcf --variant ISB/ISB.raw.snps.indels.chr$chr.union.regions.g.vcf --variant NIMHpeds/NIMHpeds.raw.snps.indels.chr$chr.union.regions.g.vcf --variant PIC/PIC.raw.snps.indels.chr$chr.g.vcf --variant RareBliss/RareBliss.raw.snps.indels.chr$chr.union.regions.final.g.vcf"
done
