Some tools for NGS studies

1. vcfSummary.py

usage: vcfSummary.py [-h] input output

Get variant- and individual-level summary from a VCF file. For example: AC, AF, missing rate, ... for variants; NVAR, Ti/Tv, ... for subjects

positional arguments:
  input     The VCF input file
  output    The prefix of output files

optional arguments:
  -h, --help  show this help message and exit

2. SelectVariants.py

usage: SelectVariants.py [-h] [--genemodel {ensembl,refSeq}] [--maf N]
                          [--splicing] [--frameshift] [--nonsynonymous]
                          [--stopgain] [--stoploss]
                          input output

Select variants based on MAF and functional categories

positional arguments:
  input     input file generated by 'table_annovar.pl'
  output    output file with all selected variants included

optional arguments:
  -h, --help      show this help message and exit
  --genemodel {ensembl,refSeq}
                        which gene model to use (default: ensembl)
  --maf N               variants with MAF > N will be excluded (default: 0.01)
  --splicing            use this flag to select splicing variants
  --frameshift          use this flag to select frameshift-indels
  --nonsynonymous       use this flag to select non-synonymous variants
  --stopgain            use this flag to select stop-gain variants
  --stoploss            use this flag to select stop-loss variants
